skopt/space/space.py

```diff
@@ -721,3 +724,108 @@ def distance(self, point_a, point_b):
             distance += dim.distance(a, b)
 
         return distance
+
+    def get_dimension(self, id):
+        """Get the Dimension-object in the search-space using either
+        its name or index as the lookup-id.
+
+        Parameters
+        ----------
+        * `id` str or int:
+            If `id` is a string then it is the name of a dimension.
+            If `id` is an integer then it is the index of a dimension.
+
+        Returns
+        -------
+        * (`dimension`, `index`, `name`) [tuple(Dimension, int, str)]:
+            Dimension-object and its index in the search-space.
+            The name is taken from the Dimension-object if not None,
+            otherwise it is 'X_n' where n is the index.
+        """
+
+        if isinstance(id, str):
+            # The id is assumed to be a string-name of a dimension,
+            # so find its index.
+            index = self.dimension_names.index(id)
+        elif isinstance(id, int):
+            # The id is assumed to be the index of a dimension.
+            index = id
+        else:
+            raise ValueError("Dimension id must be either a string with the name or an integer for the index.")
+
+        # Get the corresponding dimension-object.
+        dimension = self.dimensions[index]
+
+        # Get the name of the dimension.
+        name = dimension.name
+
+        # Use a default name for unnamed dimensions.
+        # TODO: Maybe it would be better to set this in Space.__init__?
+        if name is None:
```

Comment:
Dimensions by default do not have a name, we should only give them "fake" names in the place where we need them (like in plotting) instead of here.

Reply:
I wonder if it even makes sense to have some dimensions named and others not?

I think it is quite inelegant that some dimensions don't have names because we then have special cases elsewhere e.g. in plotting. So if we want to give them default names it might as well be done in the `Space`-class.

Let us think about what dimension-names are used for:

* With PR #590 dim-names are now used for calling the objective function.
* Dim-names are used in plotting for the names on the axes.
* Dim-names may be used to lookup a dim from the search-space object e.g. using the dict-like syntax you proposed above.
* Anything else?

Can you name a case where we don't want dimensions to have names, except for benchmark functions?

Reply:
This is really a key question! It makes the code very messy that there is the possibility for having unnamed dimensions. It requires special cases in so many places in the code. If the only reason for having unnamed dimensions is to have support for simple benchmark functions, then it would be better to drop this "feature" and force all dimensions to have names. Please explain why you have unnamed dimensions.

Reply:
Unnamed dimensions are part historical (to support the shorthand way of defining them), for small search spaces you don't need to have names, and for some problems the dimensions do not have (sensible) names. I'd be -1 on requiring all dimensions to have names (or giving them dummy names automatically).

Reply:
I am currently in the process of rewriting `plots.py` They now use the dimension-names for axis-labels. I think the most sensible way of supporting unnamed dimensions, is to return `X_i` for the i'th dimension, or something similar. I have already added an index to `Dimension` which is necessary when plotting only a sub-set of the dimensions (you'll see when I commit the code). I have just now modified the `name` property in the `Dimension`-class to return `X_index` if `self._name is None`. This also allows you to lookup unnamed dimensions using e.g. `space['X_2']` as well as named dimensions using `space['foo']`. It works fine and I think it is a reasonable solution. The only problem is that you cannot use different names for the axis in the plots. Sure it can be coded, but it would be a mess. If you want to name the axis in the plot, then they should be set when you define the `Dimension`. I think this is reasonable.